compiler/src/lnntoamm/Types.ts

```diff
@@ -197,83 +252,426 @@ class Struct extends Type {
     return this; // TODO: this right?
   }
 
-  tempConstrain(to: Type) {
+  tempConstrain(to: Type, scope: Scope) {
     // TODO: can structs have temp constraints?
-    this.constrain(to);
+    this.constrain(to, scope);
   }
 
   resetTemp() {
     // TODO: can structs have temp constraints?
   }
+
+  dupIfNotLocalInterface(): Type | null {
+    return null;
+  }
+}
+
+abstract class Has extends Type {
+  constructor(
+    name: string,
+    ast: LPNode | null,
+  ) {
+    super(name, ast);
+  }
+
+  static field(field: HasField, ty: Type): boolean {
+    // TODO: structs
+    return false;
+  }
+
+  static method(method: HasMethod, scope: Scope, ty: Type): Fn[] {
+    let fns = scope.get(method.name);
+    if (!isFnArray(fns)) {
+      return [];
+    }
+    return Fn.select(fns, method.params.map(p => p === null ? ty : p), scope);
+  }
+
+  static operator(operator: HasOperator, scope: Scope, ty: Type): Operator[] {
+    let ops: Operator[] = scope.get(operator.name);
+    // if there is no op by that name, RIP
+    if (!isOpArray(ops)) {
+      return [];
+    }
+    // filter out ops that aren't the same fixity
+    ops = ops.filter(op => op.isPrefix === operator.isPrefix);
+    if (operator.isPrefix) {
+      return ops.filter(op => op.select(scope, operator.params[0] || ty) !== []);
+    } else {
+      return ops.filter(op => op.select(scope, operator.params[0] || ty, operator.params[1] || ty) !== []);
+	  }
+  }
+
+  breakdown(): Builtin {
+    throw new Error(`cannot breakdown a Has constraint (this error should never be thrown)`);
+  }
+
+  // convenience for `Type.hasX(...).compatibleWithConstraint(ty)`
+  compatibleWithConstraint(ty: Type, scope: Scope): boolean {
+    return ty.compatibleWithConstraint(this, scope);
+  }
+
+  // convenience for `Type.hasX(...).constrain(ty)`
+  constrain(to: Type, scope: Scope) {
+    to.constrain(this, scope);
+  }
+
+  eq(that: Equalable): boolean {
+    return that instanceof Has && that.name === this.name;
```

Comment:
What if there's a field and a method of the same name? Like `@std/http`'s `Request` `body`/`body(string)`?

Reply:
should get caught by the `HasField.eq` and `HasMethod.eq` implementations - this is only for a little bit of delegation (the `instanceof Has` is redundant but it lets TS recognize that `that.name` is a valid symbol)